vsintegration/src/FSharp.Editor/CodeLens/LineLens.fs

```diff
@@ -0,0 +1,697 @@
+    let mutable trackingSpans = Dictionary<_, ResizeArray<_>>()
+    /// Text view for accessing the adornment layer.
+    let mutable view: IWpfTextView = view
+    /// The line lens layer for adding and removing adornments.
+    let mutable lineLensLayer: IAdornmentLayer = view.GetAdornmentLayer "LineLens"
+    /// Tracks the recent first + last visible line numbers for adornment layout logic.
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    /// Tracks the adornments on the layer.
+    let addedAdornments = HashSet()
+    /// Cancellation token source for the layout changed event. Needed to abort previous async-work.
+    let mutable layoutChangedCts = new CancellationTokenSource()
+
+    /// Tracks the last used buffer snapshot, should be preferred used in combination with mutex.
+    let mutable currentBufferSnapshot = null
+
+    /// Helper method which returns the start line number of a tracking span
+    let getTrackingSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        snapshot.GetLineNumberFromPosition(trackingSpan.GetStartPoint(snapshot).Position)
+
+    /// Helper method which returns the start line number of a tracking span
+    let tryGetTSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
+        let pos = trackingSpan.GetStartPoint(snapshot).Position
+        if snapshot.Length - 1 < pos then None
+        else pos |> snapshot.GetLineNumberFromPosition |> Some
+    
+    let updateTrackingSpansFast (snapshot:ITextSnapshot) lineNumber =
+        if lineNumber |> trackingSpans.ContainsKey then
+            let currentTrackingSpans = ResizeArray(trackingSpans.[lineNumber])
+            for trackingSpan in currentTrackingSpans do
+                let newLineOption = tryGetTSpanStartLine snapshot trackingSpan
+                match newLineOption with 
+                | None -> ()
+                | Some newLine ->
+                    if newLine <> lineNumber then
+                        trackingSpans.[lineNumber].Remove(trackingSpan) |> ignore
+                        if trackingSpans.[lineNumber].Count = 0 then
+                            trackingSpans.Remove lineNumber |> ignore
+                        if newLine |> trackingSpans.ContainsKey |> not then
+                            trackingSpans.[newLine] <- ResizeArray()
+                        trackingSpans.[newLine].Add trackingSpan
+                        if newLine < recentFirstVsblLineNmbr || newLine > recentLastVsblLineNmbr then
+                            if uiElements.ContainsKey trackingSpan then 
+                                let mutable element = uiElements.[trackingSpan]
+                                element.Visibility <- Visibility.Hidden
+                        // tagsChangedEvent.Trigger(self, SnapshotSpanEventArgs(trackingSpan.GetSpan snapshot)) // This results in super annoying blinking
+
+    let createDefaultStackPanel () = 
+        new Grid(Visibility = Visibility.Hidden)
```

Comment:
Might want to rename to something like `createDefaultContainer` or `createDefaultContentControl`, since it's not a stackpanel anymore (hooray!)